python/ssm_patch_compliance.py

```diff
@@ -0,0 +1,142 @@
+# tag and for each instance it will insert a compliant/non compliant evaluation into AWS config.
+# IAM role running this lambda should be able to call ec2:describe , ssm:list_compliance_items and config:putEvaluations
+
+import json
+import boto3
+
+
+
+
+APPLICABLE_RESOURCES = ["AWS::EC2::Instance"]
+
+
+def validate_patch_baseline(instanceid):
+    client = boto3.client('ssm')
+    filters = [
+		{'Key':'ComplianceType','Values':['Patch'],'Type':'EQUAL'},
+		{'Key':'Status','Values':['NON_COMPLIANT'],'Type':'EQUAL'}
+	     ]
+    response = client.list_compliance_items(Filters=filters,ResourceIds=[instanceid],ResourceTypes=['ManagedInstance'])
+    print(response)
+    if len(response['ComplianceItems']) > 0:
+        #link_to_ssm = '<a href="https://console.aws.amazon.com/systems-manager/managed-instances/' + instanceid + '/configurationCompliance>"'
+        security_critical = ''
+        
+        for item in response['ComplianceItems']:
+            
+            if 'Classification' in item['Details']:
+                if item['Details']['Classification'] == 'Security':
+                    #security_critical = security_critical + item['Title'] + ":"
+                    security_critical = " Several pending security patches."
+        violation="NON_COMPLIANT: " + instanceid  + ":" + security_critical
+        print(violation)
+        return violation
+
+    response = client.describe_instance_patch_states(InstanceIds=[instanceid])
+    
+    if len(response['InstancePatchStates']) == 0:
+        violation = 'NON_MANAGED'
+        return violation
+    
+    
+    return None
+
+
+def evaluate_compliance(configuration_item,instanceid):
+    if configuration_item["resourceType"] not in APPLICABLE_RESOURCES:
+        return {
+            "compliance_type": "NOT_APPLICABLE",
+            "annotation": "The rule doesn't apply to resources of type " +
+            configuration_item["resourceType"] + "."
+        }
+
+
+    if configuration_item['configurationItemStatus'] == "ResourceDeleted":
+        return {
+            "compliance_type": "NOT_APPLICABLE",
+            "annotation": "The configurationItem was deleted " +
+                          "and therefore cannot be validated"
+        }
+
+
+    violation = validate_patch_baseline(instanceid)
+
+    if violation == 'NON_MANAGED':
+        return {
+            "compliance_type" : "NON_COMPLIANT",
+            "annotation" : instanceid + " is not managed by SSM"
+        }
+
+    if violation:
+            return {
+                "compliance_type": "NON_COMPLIANT",
+                "annotation": violation
+            }
+
+    return {
+        "compliance_type": "COMPLIANT",
+        "annotation": "This resource is compliant with the rule."
+    }
+
+
+
+def lambda_handler(event, context):
+
+    print(event)
+    invoking_event = json.loads(event["invokingEvent"])
+    #configuration_item = invoking_event["configurationItem"]
```

Comment:
Why? This rule should be configuration changes triggered, as far I understand the intend.

Reply:
At the same time there may be occasions when we want to run the config rule manually to ensure that all instances are patch compliant ie in event of some major security vulnerability announcement. But I think this gives a good foundation for users to customize as needed.